library/alloc/src/vec/mod.rs

```diff
@@ -2274,6 +2274,26 @@ impl<T: Clone, A: Allocator> Vec<T, A> {
     }
 }
 
+impl<T, A: Allocator, const N: usize> Vec<[T; N], A> {
+    #[unstable(feature = "slice_flatten", issue = "none")]
+    pub fn into_flattened(self) -> Vec<T, A> {
+        let (ptr, len, cap, alloc) = self.into_raw_parts_with_alloc();
+        let (new_len, new_cap) = if mem::size_of::<T>() == 0 {
+            (len.checked_mul(N).expect("vec len overflow"), usize::MAX)
+        } else {
+            // SAFETY:
+            // - `cap * N` cannot overflow because the allocation is already in
+            // the address space.
+            // - Each `[T; N]` has `N` valid elements, so there are `len * N`
+            // valid elements in the allocation.
+            // - `len` <= `cap`, so `len * N` <= `cap * N`.
+            unsafe { (len.unchecked_mul(N), cap.unchecked_mul(N)) }
+        };
+        // SAFETY: `new_len` and `new_cap` are valid, see above.
```

Comment:
suggestion: mention the alignment equality too, since this is only sound because the alignment of an array is the same as the alignment of the elements.  (Doing this with SIMD wouldn't be sound, for example.)